File: backend/app/Support/LateMinutes.php
<?php

namespace App\Support;

use App\Models\AttendanceSession;
use Carbon\CarbonInterface;
use Illuminate\Support\Carbon;

/**
 * دقائق تأخير الطالب — محسوبةً من بداية الدوام لا من لحظة فتح الجلسة.
 *
 * المرجع `shifts.starts_at` ثابتٌ ومعروف قبل أن يفتح أحدٌ شيئاً، فالرقم قابل للمقارنة
 * بين حلقةٍ وأخرى وبين جهازٍ وآخر. ولو قِيس من لحظة الفتح لصار «تأخيرُ الطالب» تابعاً
 * لتأخير أستاذه في فتح الجلسة، ولاختلف الرقمُ نفسه باختلاف من فتح.
 *
 * القرار مسجَّل في PHASE-5-STAGES.MD §0 البند 1، ونظيره في العميل
 * (packages/mousqe_core) يطبّق **نفس** هذه الصيغة ليعرض الأستاذ الرقم فوراً أوف-لاين
 * قبل أن يؤكّده الخادم.
 */
class LateMinutes
{
    /**
     * الدقائق، أو null حين لا مرجع يُقاس عليه.
     *
     * ثلاث حالات تعود null قصداً:
     * 1. لا دوام للحلقة (بيانات ناقصة) — لا مرجع أصلاً.
     * 2. زمن التسجيل من **يوم غير يوم الجلسة**: التفقّد الرجعي لتاريخٍ ماضٍ يُسجَّل اليوم،
     *    فالفرق بينه وبين بداية دوام ذلك اليوم أيامٌ لا دقائق. رقمٌ بلا معنى، وتركُه
     *    فارغاً أصدق من ملئه بعددٍ خيالي.
     * 3. الطالب وصل قبل بداية الدوام أو معها ⇒ صفر لا سالب.
     */
    public static function forSession(AttendanceSession $session, ?CarbonInterface $recordedAt = null): ?int
    {
        $startsAt = $session->courseCircle?->shift?->starts_at;

        if (blank($startsAt)) {
            return null;
        }

        $recordedAt = $recordedAt ?? now();

        $sessionDate = $session->session_date instanceof CarbonInterface
            ? $session->session_date->toDateString()
            : (string) $session->session_date;

        if ($recordedAt->toDateString() !== $sessionDate) {
            return null;
        }

        $start = Carbon::parse($sessionDate)->setTimeFromTimeString((string) $startsAt);

        return max(0, (int) floor($start->diffInMinutes($recordedAt, absolute: false)));
    }

    /**
     * الدقائق بعد خصم فترة السماح المضبوطة في إعدادات المعهد.
     *
     * فترة السماح صفرٌ افتراضاً، فيكون الرقمُ حرفياً «كم دقيقة بعد بداية الدوام». معهدٌ
     * يريد تسامحاً يضبطها فيصير الرقم «كم دقيقة تأخيرٍ غير مقبول».
     */
    public static function afterGrace(?int $minutes, int $graceMinutes): ?int
    {
        if ($minutes === null) {
            return null;
        }

        return max(0, $minutes - $graceMinutes);
    }
}

File: packages/mousqe_core/lib/src/support/late_minutes.dart
/// نظير `backend/app/Support/LateMinutes.php` حرفياً — دقائق التأخير محسوبةً من
/// بداية الدوام لا من لحظة فتح الجلسة، فيعرض العميل الرقم فوراً أوف-لاين بنفس
/// الصيغة التي يؤكّدها الخادم لاحقاً. القرار في PHASE-5-STAGES.MD §0 البند 1.
class LateMinutes {
  const LateMinutes._();

  /// الدقائق، أو `null` حين لا مرجع يُقاس عليه.
  ///
  /// ثلاث حالات تعود `null` قصداً:
  /// 1. لا دوام للحلقة (`shiftStartsAt` فارغ) — لا مرجع أصلاً.
  /// 2. `recordedAt` من **يومٍ غير يوم الجلسة**: تفقّدٌ رجعي لتاريخ ماضٍ يُسجَّل
  ///    اليوم، فالفرق أيامٌ لا دقائق.
  /// 3. الطالب وصل قبل بداية الدوام أو معها ⇒ صفر لا سالب (تُرجَع كـ 0 لا null).
  static int? forSession({
    required String? shiftStartsAt,
    required DateTime sessionDate,
    DateTime? recordedAt,
  }) {
    if (shiftStartsAt == null || shiftStartsAt.isEmpty) {
      return null;
    }

    final effectiveRecordedAt = recordedAt ?? DateTime.now();

    if (!_isSameDate(effectiveRecordedAt, sessionDate)) {
      return null;
    }

    final start = _combine(sessionDate, shiftStartsAt);
    final diffMinutes = effectiveRecordedAt.difference(start).inMinutes;

    return diffMinutes < 0 ? 0 : diffMinutes;
  }

  /// الدقائق بعد خصم فترة السماح المضبوطة في إعدادات المعهد.
  static int? afterGrace(int? minutes, int graceMinutes) {
    if (minutes == null) {
      return null;
    }

    final result = minutes - graceMinutes;
    return result < 0 ? 0 : result;
  }

  static bool _isSameDate(DateTime a, DateTime b) {
    return a.year == b.year && a.month == b.month && a.day == b.day;
  }

  static DateTime _combine(DateTime date, String timeOfDay) {
    final parts = timeOfDay.split(':');
    final hour = int.parse(parts[0]);
    final minute = parts.length > 1 ? int.parse(parts[1]) : 0;
    final second = parts.length > 2 ? int.parse(parts[2]) : 0;

    return DateTime(date.year, date.month, date.day, hour, minute, second);
  }
}
